### ISSUE TITLE
Accuracy error after converting

### ISSUE THREAD

**Issue by IvanJuraga** [PINTO0309/openvino2tensorflow#8]
Greetings,

I'm trying to convert retinaface with mobilenet0.25 as the backbone and I'm getting error in predictions so I'm wondering is that the same problem as with mobilenet v3 because I saw that you managed to convert retinaface but with resnet50 as the backbone. Any information would be appriciated.

**Comment by PINTO0309** [comment]
If you can provide a model I can verify.
Also, please let me know the exact versions of the various tools you used.

**Comment by IvanJuraga** [comment]
```
absl-py==0.11.0
astor==0.8.1
astunparse==1.6.3
cached-property==1.5.2
cachetools==4.1.1
certifi==2020.11.8
chardet==3.0.4
dataclasses==0.8
future==0.18.2
gast==0.3.3
google-auth==1.23.0
google-auth-oauthlib==0.4.2
google-pasta==0.2.0
grpcio==1.33.2
h5py==2.10.0
idna==2.10
importlib-metadata==3.1.1
Keras-Applications==1.0.8
Keras-Preprocessing==1.1.2
Markdown==3.3.3
numpy==1.18.5
oauthlib==3.1.0
onnx==1.8.0
-e git+https://github.com/onnx/onnx-tensorflow.git@80d28066b42f2adff5190a4f8051d723e44ca0a9#egg=onnx_tf
onnxruntime==1.6.0
opencv-python==4.4.0.46
opt-einsum==3.3.0
Pillow==8.0.1
protobuf==3.14.0
pyasn1==0.4.8
pyasn1-modules==0.2.8
PyYAML==5.3.1
requests==2.25.0
requests-oauthlib==1.3.0
rsa==4.6
scipy==1.4.1
six==1.15.0
tensorboard==2.4.0
tensorboard-plugin-wit==1.7.0
tensorflow-addons==0.11.2
tensorflow-estimator==2.3.0
tensorflow-gpu==2.3.1
termcolor==1.1.0
torch==1.7.1
torchvision==0.8.2
typeguard==2.10.0
typing-extensions==3.7.4.3
urllib3==1.26.2
Werkzeug==1.0.1
wrapt==1.12.1
zipp==3.4.0
```
This is the env in which I do all of my conversions. The pipeline I followed was pytorch -->onnx-->onnx_simplify --openvino -->tf -->tflite

and about the model I verifyed that my model is correct up until the openvino2tf part. Do you want just the openvino file or?

**Comment by PINTO0309** [comment]
Let me ask you a few more questions.
1. Was it Float32 or INT8 that caused the problem?
2. Can you share the command you ran?

>Do you want just the openvino file or?

If possible, please provide me with the ONNX and OpenVINO IR before conversion so I can look into it.

**Comment by IvanJuraga** [comment]
1.Right now I just want to get the same inference results from pytorch model and a fp32 tflite model.
2. If you mean all the conversion command then I used the pytorch to onnx conversion script which was already in the repo. and from there I used the following commands:

**Comment by PINTO0309** [comment]
## Environment
1. tf-nightly==2.5.0-dev20201227
2. openvino2tensorflow==1.4.4

## Conversion result
I have performed the conversion based on the ONNX file you provided, and it appears to be structurally sound. I have been working on bug fixes for openvino2tensorflow for the last few times, so the structure may be different from your tflite. If you still have problems with the inference results, the behavior of **`ResizeNearestNeighbor`** is suspicious.
- Converted tflite float32
**https://drive.google.com/file/d/1W5qLnLw2SFlYaEbn8Me6a19wa-HN-qgv/view?usp=sharing**

Is it possible to share a program for testing inference?

**Comment by IvanJuraga** [comment]
Hi I tried using the versions you recommended but I get the same accuracy loss. I shared you a sample image together with the onnx output / desired output and the output I get when I use the tflite converted model. If you suspect the ResizeNN to be guilty of this loss increase is there any known way to correct it.

pack: https://file.io/s4zqzFXEUajJ

**Comment by PINTO0309** [comment]
@IvanJuraga 
When I go to the URL you provided, the file is not found.

**Comment by IvanJuraga** [comment]
Oh looks like file.io site deletes files after a certain time passed. Here you go a fresh link
https://file.io/n83fkllIcOwf

**Comment by PINTO0309** [comment]
I found the cause. This is the very problem that occurs with **`ResizeNearestNeighbor`**. All detection results will be shifted to the upper left. Sorry, This is a problem with the TFLite framework itself, so it is difficult to resolve at the moment.

**Comment by PINTO0309** [comment]
I'll do some research.

**Comment by IvanJuraga** [comment]
Well first thank you for scoping in on the problem of the model but I have the same shift to upper left problem in the frozen graph so the problem already occurs in the TF framework. So is there a way in which I can maybe manually fix this problem or is it something we/I need on TF to fix itself?

**Comment by PINTO0309** [comment]
I will investigate that as well, but can I get the ONNX file again first?

**Comment by IvanJuraga** [comment]
Here you go 
https://file.io/lK7Ko4zPP5Ni
I added both the converted onnx file as well the simplified onnx to save you the hassle of doing it.

**Comment by PINTO0309** [comment]
ONNX -> onnx-tensorflow -> saved_mode -> tflite
I'm starting to wonder about the behavior of the openvino optimizer.
![Screenshot 2021-01-11 20:41:50](https://user-images.githubusercontent.com/33194443/104178331-7bbb0080-544d-11eb-8e27-2fd24dde9779.png)

**Comment by IvanJuraga** [comment]
Yeah the landmarks just need to be adjusted but that is good. I already explored that route and got the results same as you did but I wanted to see and try to get the desired tflite model using the route onnx -> openvino -> tf -> tflite. One of the reasons is that with the second route you get a much "cleaner" model and curiosity

**Comment by IvanJuraga** [comment]
The problem is that I pin pointed the error to happen while converting from openvino to tensorflow thus why I posted this issue on your project and not somewhere else to see if you had any prior expirience.

Also maybe an out of nowhere question. This tflite model you made is it a fp32 or an all uint8 model? Because I'm also facing a problem of quantize the output of the bboxes

**Comment by PINTO0309** [comment]
Finally, I understood. It seems that there is a problem with the conversion logic of my openvino2tensorflow. In particular, there seems to be a problem with the GroupConvolution conversion logic. However, it is clear that if I fix my logic correctly, it produces a model that is not clean. This cannot be avoided because TFLite does not support GroupConvolution.
![Screenshot 2021-01-11 21:00:49](https://user-images.githubusercontent.com/33194443/104180107-1c122480-5450-11eb-8e4f-4111f01cde0f.png)

**Comment by IvanJuraga** [comment]
So is it possible to fix the logic just to try is the new model we get from the "fixed" logic giving the same output as the starting model. At that point I would just like to check out this route and am not really worried about how clean the model is.

Also would appreciate if you could comment about the quantization problem I presented in the last comment.

**Comment by PINTO0309** [comment]
>This tflite model you made is it a fp32 or an all uint8 model?

It is fp32.

openvino2tensorflow will perform INT8 quantization only when set to "--output_integer_quant_tflite True" or "--output_full_integer_quant_tflite True". I know that the accuracy of heatmap-based models deteriorates significantly when INT8 quantization is used. Examples are Posenet and OpenPose. I am not sure yet if this is the same problem you are facing this time.

**Comment by IvanJuraga** [comment]
the significant error you are talking about are you refering to all the outputs or just some? In my case only with the full INT8 quantization only the bbox results are worse while the confs and the landmarks look almost the same.

Also is ti possible to get a patch in which I can fix the logic of the openvino2tensorflow lib so I can test it out on my model or can you convert the model I provided to you today and send me back the converted .pb file so I can check it out

**Comment by PINTO0309** [comment]
Thank you for providing very interesting insights.

>the significant error you are talking about are you refering to all the outputs or just some?

The extreme degradation of accuracy that I have experienced has been with Landmark. So the report of the phenomenon you have just provided is new to me.

>Also is ti possible to get a patch in which I can fix the logic of the openvino2tensorflow lib so I can test it out on my model or can you convert the model I provided to you today and send me back the converted .pb file so I can check it out

Thank you very much for your cooperation. I have temporarily modified the program at hand to be able to convert GoupConvolution, but it still looks like there is a problem. I'll try a few more things, so please wait a bit. :smiley:

**Comment by IvanJuraga** [comment]
I suspect that the problem with the INT8 quantization of the output node atleast in my case that the range of values I get from the fp32 bbox output is too wide for tflite to accuratly represent it in a INT8 range so  the part it should focus more is not a accuratly represent so I will try and hardcode the range of the values in hopes to get a better quantization results. If I manage to get a correct full int8 model I will update you.

On the topic of the patch I asked for I understand that debugging takes time and I will patiently wait until you provide me with the patch.

**Comment by PINTO0309** [comment]
It worked.
![Screenshot 2021-01-11 23:02:10](https://user-images.githubusercontent.com/33194443/104191620-1113c000-5461-11eb-941c-bb7f682532a7.png)

**Comment by IvanJuraga** [comment]
great :D

**Comment by PINTO0309** [comment]
The problem is that I'm still not convinced that simply inserting a PAD before GoupConvolution improves the accuracy. It's a very rough fix specific to the model you provided, so I'll tweak it some more. :sweat_smile:

**Comment by IvanJuraga** [comment]
I understand. I thought so that this is a really model specific fix and I'll wait until you are more confident that the fix you applied to the model is a more general solution. If you could atleast send the .pb model so I can play around with it and quantization processes I would be grateful.

**Comment by PINTO0309** [comment]
This is a .pb created on a trial basis. Please feel free to use it.
**https://drive.google.com/file/d/1-lzSB082jGcgLRxyERZeFHarC5EzyWwe/view?usp=sharing**

**Comment by PINTO0309** [comment]
The problem has been fixed and imported into the tool, which is now closed.
